File: js/src/noise.ts
/**
 * Render randomness as data — a port of `loudkit.models.noise`.
 *
 * The flow prior and the vocoder excitation are *inputs* that happen to be
 * random. They come from the Philox counter, so the same seed produces the
 * same bytes on every backend, and a cross-backend render comparison measures
 * arithmetic rather than RNG plumbing.
 */

import { uniforms } from "./rng.js";

export function gaussianField(
  seed: bigint,
  stream: number,
  rows: number,
  cols: number
): Float32Array {
  const u1 = uniforms(seed, stream, 0, rows, cols);
  const u2 = uniforms(seed, stream + 1, 0, rows, cols);
  const out = new Float32Array(rows * cols);
  for (let i = 0; i < rows * cols; i++) {
    out[i] = Math.sqrt(-2.0 * Math.log(u1[i])) * Math.cos(2.0 * Math.PI * u2[i]);
  }
  return out;
}

export function symmetricUniforms(
  seed: bigint,
  stream: number,
  n: number,
  halfWidth: number
): Float32Array {
  const u = uniforms(seed, stream, 0, 1, n);
  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    out[i] = (u[i] * 2.0 - 1.0) * halfWidth;
  }
  return out;
}

File: rust/src/noise.rs
//! Render randomness as Philox data — a bit-parity port of
//! `loudkit.models.noise`.

use crate::rng;

/// rows*cols standard-normal field.
pub fn gaussian_field(seed: u64, stream: u32, rows: usize, cols: usize) -> Vec<f32> {
    let u1 = rng::uniforms(seed, stream, 0, rows, cols);
    let u2 = rng::uniforms(seed, stream.wrapping_add(1), 0, rows, cols);
    u1.iter()
        .zip(u2.iter())
        .map(|(a, b)| ((-2.0 * a.ln()).sqrt() * (2.0 * std::f64::consts::PI * b).cos()) as f32)
        .collect()
}

/// n uniforms in (-half_width, half_width).
pub fn symmetric_uniforms(seed: u64, stream: u32, n: usize, half_width: f64) -> Vec<f32> {
    let u = rng::uniforms(seed, stream, 0, 1, n);
    u.iter()
        .map(|v| ((v * 2.0 - 1.0) * half_width) as f32)
        .collect()
}
